File: examples/ex-hello/lib/hello.ex
defmodule Hello do
  # The first argument is the path to the giftwrapped binary!
  def main([_]),       do: greet()
  def main([_, name]), do: greet(name)
  def main(_),         do: usage()

  defp greet(name \\ "world"), do: IO.puts("Hello, #{name}!")

  defp usage do
    IO.puts(:stderr, "usage: hello [name]")
    :erlang.halt(1)
  end
end

File: examples/erl-hello/src/hello.erl
-module(hello).
-export([main/1]).

%% The first argument is the path to the giftwrapped binary!
main([_])       -> greet("world");
main([_, Name]) -> greet(Name);
main(_)         -> usage().

greet(Name) ->
  io:fwrite("Hello, ~s!~n", [Name]).

usage() ->
  io:fwrite(standard_error, "usage: hello [name]~n", []),
  halt(1).
